src/services/io/useIO.ts

```diff
@@ -442,71 +447,121 @@ export const useIO = create<IOStore>((set, get) => ({
+
+      const shotcutMltXml = await generateMLT()
+      files['shotcut_project.mlt'] = fflate.strToU8(shotcutMltXml)
+
+      const videos: FFMPegVideoInput[] = []
+      const audios: FFMPegAudioInput[] = []
+
+      segments.forEach(
+        ({
+          segment,
+          prefix,
+          filePath,
+          assetUrl,
+          assetSourceType,
+          isExportableToFile,
+        }) => {
+          // we extract the base64 files
+          if (isExportableToFile) {
+            files[filePath] = [
+              // we don't compress assets since normally they already use
+              // some form of compression (PNG, JPEG, MP3, MP4 etc..)
+              base64DataUriToUint8Array(segment.assetUrl),
+              { level: 0 },
+            ]
+            assetUrl = filePath
+            assetSourceType = ClapAssetSource.PATH
+
+            if (filePath.startsWith('video/')) {
+              videos.push({
+                data: base64DataUriToUint8Array(segment.assetUrl),
+                startTimeInMs: segment.startTimeInMs,
+                endTimeInMs: segment.endTimeInMs,
+                durationInSecs: segment.assetDurationInMs / 1000,
+              })
+            }
+
+            if (
+              filePath.startsWith('music/') ||
+              filePath.startsWith('dialogue/')
+            ) {
+              audios.push({
+                data: base64DataUriToUint8Array(segment.assetUrl),
+                startTimeInMs: segment.startTimeInMs,
+                endTimeInMs: segment.endTimeInMs,
+                durationInSecs: segment.assetDurationInMs / 1000,
+              })
+            }
+          }
+
+          // segment metadata
+          files[`segments/${prefix}${segment.id}.json`] = fflate.strToU8(
+            JSON.stringify(
+              {
+                ...segment,
+                assetUrl,
+                assetSourceType,
+              },
+              null,
+              2
+            )
+          )
         }
+      )
 
-        // segment metadata
-        files[`segments/${prefix}${segment.id}.json`] = fflate.strToU8(
-          JSON.stringify(
-            {
-              ...segment,
-              assetUrl,
-              assetSourceType,
-            },
-            null,
-            2
-          )
-        )
-      }
-    )
+      const fullVideo = await createFullVideo(
+        videos,
+        audios,
+        1024,
+        576,
+        timeline.totalDurationInMs,
+        (progress, message) => {
+          task.setProgress({
+            message,
+            value: progress * 0.9,
+          })
+        }
+      )
 
-    fflate.zip(
-      files,
-      {
-        // options
-      },
-      (error, zipFile) => {
-        task.setProgress({
-          message: 'Saving to file..',
-          value: 90,
-        })
-        saveAnyFile(new Blob([zipFile]), 'my_project.zip')
-        task.success()
-      }
-    )
+      files['video/full.mp4'] = [
```

Comment:
FYI there is a dedicated menu entry "Export to MP4" in here:
https://github.com/jbilcke-hf/clapper/blob/main/src/components/toolbars/top-menu/file/index.tsx#L121-L129

We can reactivate it, and replace this `saveVideoFile()` function by your code:

https://github.com/jbilcke-hf/clapper/blob/main/src/services/io/useIO.ts#L378-L428

that way the export to .zip can stay fast & related to exporting individual assets only, and keep the generation of the mp4 separate

but it's not a blocker I can do it right after we merge your PR, that's not a problem 🙂